source/val/validate_builtins.cpp

```diff
@@ -4190,60 +4260,164 @@ spv_result_t BuiltInsValidator::ValidateMeshShadingEXTBuiltinsAtDefinition(
-                return _.diag(SPV_ERROR_INVALID_DATA, &inst)
-                       << _.VkErrorID(vuid) << "According to the "
-                       << spvLogStringForEnv(_.context()->target_env)
-                       << " spec BuiltIn "
-                       << _.grammar().lookupOperandName(
-                              SPV_OPERAND_TYPE_BUILT_IN,
-                              (uint32_t)decoration.builtin())
-                       << " variable needs to be a 2-component 32-bit int "
-                          "array."
-                       << message;
-              })) {
+    for (const uint32_t entry_point : _.entry_points()) {
+      const auto* modes = _.GetExecutionModes(entry_point);
+      uint32_t maxOutputPrimitives = _.GetOutputPrimitivesEXT(entry_point);
+      uint32_t underlying_type = 0;
+      if (spv_result_t error =
+              GetUnderlyingType(_, decoration, inst, &underlying_type)) {
         return error;
       }
-    }
-    if (builtin == spv::BuiltIn::PrimitiveTriangleIndicesEXT) {
-      if (spv_result_t error = ValidateArrayedI32Vec(
-              decoration, inst, 3,
-              [this, &inst, &decoration,
-               &vuid](const std::string& message) -> spv_result_t {
-                return _.diag(SPV_ERROR_INVALID_DATA, &inst)
-                       << _.VkErrorID(vuid) << "According to the "
-                       << spvLogStringForEnv(_.context()->target_env)
-                       << " spec BuiltIn "
-                       << _.grammar().lookupOperandName(
-                              SPV_OPERAND_TYPE_BUILT_IN,
-                              (uint32_t)decoration.builtin())
-                       << " variable needs to be a 3-component 32-bit int "
-                          "array."
-                       << message;
-              })) {
-        return error;
+
+      uint32_t primitiveArrayDim = 0;
+      // Strip the array, if present.
+      if (_.GetIdOpcode(underlying_type) == spv::Op::OpTypeArray) {
+        underlying_type = _.FindDef(underlying_type)->word(3u);
+        if (_.GetIdOpcode(underlying_type) == spv::Op::OpConstant) {
+          assert(_.GetIdOpcode(_.FindDef(underlying_type)->word(1)) ==
+                 spv::Op::OpTypeInt);
+          primitiveArrayDim = _.FindDef(underlying_type)->word(3);
+        }
+      }
+      switch (builtin) {
+        case spv::BuiltIn::PrimitivePointIndicesEXT:
+          if (!modes || !modes->count(spv::ExecutionMode::OutputPoints)) {
+            return _.diag(SPV_ERROR_INVALID_DATA, &inst)
+                   << _.VkErrorID(7042)
+                   << "The PrimitivePointIndicesEXT decoration must be used "
+                      "with "
+                      "the OutputPoints Execution Mode. ";
+          }
+          if (primitiveArrayDim && primitiveArrayDim != maxOutputPrimitives) {
+            return _.diag(SPV_ERROR_INVALID_DATA, &inst)
+                   << _.VkErrorID(7046)
```

Comment:
@spencer-lunarg VUID-PrimitivePointIndicesEXT-PrimitivePointIndicesEXT-07046 is here